File: sdk-android-library/src/main/java/network/xyo/ble/bluetooth/ScanResultLegacy.java
package network.xyo.ble.bluetooth;

import android.bluetooth.BluetoothDevice;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Objects;

// This is a copy of ScanResult from Android Marshmallow+ to be used with Android versions earlier than Marshmallow

public final class ScanResultLegacy implements Parcelable {
    public static final Creator<ScanResultLegacy> CREATOR = new Creator<ScanResultLegacy>() {
        @NonNull
        @Override
        public ScanResultLegacy createFromParcel(@NonNull Parcel source) {
            return new ScanResultLegacy(source);
        }

        @NonNull
        @Override
        public ScanResultLegacy[] newArray(int size) {
            return new ScanResultLegacy[size];
        }
    };
    // Remote bluetooth device.
    private BluetoothDevice mDevice;
    // Scan record, including advertising data and scan response data.
    @Nullable
    private ScanRecordLegacy mScanRecord;
    // Received signal strength.
    private int mRssi;
    // Device timestamp when the result was last seen.
    private long mTimestampNanos;

    /**
     * Constructor of scan result.
     *
     * @param device         Remote bluetooth device that is found.
     * @param scanRecord     Scan record including both advertising data and scan response data.
     * @param rssi           Received signal strength.
     * @param timestampNanos Device timestamp when the scan result was observed.
     */
    public ScanResultLegacy(BluetoothDevice device, @Nullable ScanRecordLegacy scanRecord, int rssi,
                            long timestampNanos) {
        mDevice = device;
        mScanRecord = scanRecord;
        mRssi = rssi;
        mTimestampNanos = timestampNanos;
    }

    private ScanResultLegacy(@NonNull Parcel in) {
        readFromParcel(in);
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        if (mDevice != null) {
            dest.writeInt(1);
            mDevice.writeToParcel(dest, flags);
        } else {
            dest.writeInt(0);
        }
        if (mScanRecord != null) {
            dest.writeInt(1);
            dest.writeByteArray(mScanRecord.getBytes());
        } else {
            dest.writeInt(0);
        }
        dest.writeInt(mRssi);
        dest.writeLong(mTimestampNanos);
    }

    private void readFromParcel(@NonNull Parcel in) {
        if (in.readInt() == 1) {
            mDevice = BluetoothDevice.CREATOR.createFromParcel(in);
        }
        if (in.readInt() == 1) {
            mScanRecord = ScanRecordLegacy.parseFromBytes(in.createByteArray());
        }
        mRssi = in.readInt();
        mTimestampNanos = in.readLong();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    /**
     * Returns the remote bluetooth device identified by the bluetooth device address.
     */
    public BluetoothDevice getDevice() {
        return mDevice;
    }

    /**
     * Returns the scan record, which is a combination of advertisement and scan response.
     */
    @Nullable
    public ScanRecordLegacy getScanRecord() {
        return mScanRecord;
    }

    /**
     * Returns the received signal strength in dBm. The valid range is [-127, 127].
     */
    public int getRssi() {
        return mRssi;
    }

    /**
     * Returns timestamp since boot when the scan record was observed.
     */
    public long getTimestampNanos() {
        return mTimestampNanos;
    }

    @Override
    public int hashCode() {
        return Objects.hash(mDevice, mRssi, mScanRecord, mTimestampNanos);
    }

    @Override
    public boolean equals(Object obj) {

        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }

        if (this.equals(obj)) {
            return true;
        }

        ScanResultLegacy other = (ScanResultLegacy) obj;
        return Objects.equals(mDevice, other.mDevice) && (mRssi == other.mRssi) &&
                Objects.equals(mScanRecord, other.mScanRecord)
                && (mTimestampNanos == other.mTimestampNanos);
    }

    @NonNull
    @Override
    public String toString() {
        return "ScanResultLegacy{" + "mDevice=" + mDevice + ", mScanRecord="
                + Objects.toString(mScanRecord) + ", mRssi=" + mRssi + ", mTimestampNanos="
                + mTimestampNanos + '}';
    }

}

File: ble-android-library/src/main/kotlin/network/xyo/ble/bluetooth/ScanResultLegacy.kt
package network.xyo.ble.bluetooth

import android.bluetooth.BluetoothDevice
import android.os.Parcel
import android.os.Parcelable
import org.jetbrains.annotations.NotNull

import java.util.Objects

import org.jetbrains.annotations.Nullable

// This is a copy of ScanResult from Android Marshmallow+ to be used with Android versions earlier than Marshmallow

class ScanResultLegacy : Parcelable {
    // Remote bluetooth device.
    /**
     * Returns the remote bluetooth device identified by the bluetooth device address.
     */
    var device: BluetoothDevice? = null
        private set
    // Scan record, including advertising data and scan response data.
    /**
     * Returns the scan record, which is a combination of advertisement and scan response.
     */
    @Nullable
    @get:Nullable
    var scanRecord: ScanRecordLegacy? = null
        private set
    // Received signal strength.
    /**
     * Returns the received signal strength in dBm. The valid range is [-127, 127].
     */
    var rssi: Int = 0
        private set
    // Device timestamp when the result was last seen.
    /**
     * Returns timestamp since boot when the scan record was observed.
     */
    var timestampNanos: Long = 0
        private set

    /**
     * Constructor of scan result.
     *
     * @param device         Remote bluetooth device that is found.
     * @param scanRecord     Scan record including both advertising data and scan response data.
     * @param rssi           Received signal strength.
     * @param timestampNanos Device timestamp when the scan result was observed.
     */
    constructor(device: BluetoothDevice, @Nullable scanRecord: ScanRecordLegacy, rssi: Int,
                timestampNanos: Long) {
        this.device = device
        this.scanRecord = scanRecord
        this.rssi = rssi
        this.timestampNanos = timestampNanos
    }

    private constructor(@NotNull `in`: Parcel) {
        readFromParcel(`in`)
    }

    override fun writeToParcel(@NotNull dest: Parcel, flags: Int) {
        if (device != null) {
            dest.writeInt(1)
            device?.writeToParcel(dest, flags)
        } else {
            dest.writeInt(0)
        }
        if (scanRecord != null) {
            dest.writeInt(1)
            dest.writeByteArray(scanRecord!!.bytes)
        } else {
            dest.writeInt(0)
        }
        dest.writeInt(rssi)
        dest.writeLong(timestampNanos)
    }

    private fun readFromParcel(@NotNull parcel: Parcel) {
        if (parcel.readInt() == 1) {
            device = BluetoothDevice.CREATOR.createFromParcel(parcel)
        }
        if (parcel.readInt() == 1) {
            scanRecord = ScanRecordLegacy.parseFromBytes(parcel.createByteArray())
        }
        rssi = parcel.readInt()
        timestampNanos = parcel.readLong()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun hashCode(): Int {
        return Objects.hash(device, rssi, scanRecord, timestampNanos)
    }

    override fun equals(other: Any?): Boolean {

        if (other == null || javaClass != other.javaClass) {
            return false
        }

        if (other is ScanResultLegacy) {
            return (device == other.device && rssi == other.rssi &&
                    scanRecord == other.scanRecord
                    && timestampNanos == other.timestampNanos)
        }
        return false
    }

    @NotNull
    override fun toString(): String {
        return ("ScanResultLegacy{" + "mDevice=" + device + ", mScanRecord="
                + Objects.toString(scanRecord) + ", mRssi=" + rssi + ", mTimestampNanos="
                + timestampNanos + '}'.toString())
    }

    companion object {

        @JvmField
        val CREATOR: Parcelable.Creator<ScanResultLegacy> = object : Parcelable.Creator<ScanResultLegacy> {
            @NotNull
            override fun createFromParcel(@NotNull source: Parcel): ScanResultLegacy {
                return ScanResultLegacy(source)
            }

            @NotNull
            override fun newArray(size: Int): Array<ScanResultLegacy?> {
                return arrayOfNulls(size)
            }
        }
    }

}
